### PR TITLE
drivers: sensor: icm42670: Add I2C bus support

### PR DESCRIPTION
**Pull request by LucasTambor**
Add bus interface so the driver can support both
SPI and I2C bus.

### CODE REVIEW

File: dts/bindings/sensor/invensense,icm42670-i2c.yaml [hunk 1]
```diff
@@ -0,0 +1,80 @@
+# Copyright (c) 2022 Esco Medical ApS
+# Copyright (c) 2020 TDK Invensense
+# SPDX-License-Identifier: Apache-2.0
+
+description: ICM-42670 motion tracking device
+
+compatible: "invensense,icm42670"
+
+include: [sensor-device.yaml, i2c-device.yaml]
```
Comment (jeppenodgaard) [inline dts/bindings/sensor/invensense,icm42670-i2c.yaml:9]: A common yaml file e.g. `icm42670-common.yaml` should be created with all the common properties and then included here.

File: tests/drivers/build_all/sensor/i2c.dtsi [hunk 2]
```diff
@@ -142,6 +142,16 @@ test_i2c_mpu9250: mpu9250@12 {
 	accel-dlpf = "5.05";
 };
 
+test_i2c_icm42670: icm42670@18 {
```
Comment (jeppenodgaard) [inline tests/drivers/build_all/sensor/i2c.dtsi:145]: New tests should be added in the bottom with an incremental address.

**Review by jeppenodgaard — CHANGES_REQUESTED** [review]
Consider `clang-format` on `icm42670.c` to remove no longer needed newlines and improve alignment. Maybe on more files.

File: .clang-format [hunk 3]
```diff
@@ -92,3 +92,4 @@ UseTab: ForContinuationAndIndentation
 WhitespaceSensitiveMacros:
   - STRINGIFY
   - Z_STRINGIFY
+AlignEscapedNewlines: Right
```
Comment (teburd) [inline .clang-format:95]: This needs to be in a separate commit

File: drivers/sensor/tdk/icm42670/icm42670_i2c.c [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+ * Bus-specific functionality for ICM42670 accessed via I2C.
+ */
+
+#include "icm42670.h"
+#include "icm42670_reg.h"
+
+#if ICM42670_BUS_I2C
+static int icm42670_bus_check_i2c(const union icm42670_bus *bus)
+{
+	return device_is_ready(bus->i2c.bus) ? 0 : -ENODEV;
```
Comment (teburd) [inline drivers/sensor/tdk/icm42670/icm42670_i2c.c:17]: i2c_is_ready_dt(bus->i2c) can be used here

**Comment by teburd** [comment]
I'd note too that TDK has a PR using their own code for this device as a PR #71374

File: .clang-format (hunk 3 above)
Comment (LucasTambor) [reply .clang-format]: My bad, this was just a test, not supposed to be in this PR haha

**Comment by LucasTambor** [comment]
> I'd note too that TDK has a PR using their own code for this device as a PR #71374

Nice, I wasn't aware of this PR. Should I just close this one?

**Comment by teburd** [comment]
> > I'd note too that TDK has a PR using their own code for this device as a PR #71374
> 
> Nice, I wasn't aware of this PR. Should I just close this one? @teburd 

I would spend some time reviewing it to ensure it does what you want and expect, if so please do comment and approve it (even if its not a green checkmark!). That PR is waiting on hal_tdk and reviews.

**Review by marekmatej — DISMISSED** [review]

**Comment by LucasTambor** [comment]
Folks. It's been quite a while and it seems that #71374 may not be merged anytime soon. I propose we merge this commit and as soon #71374 gets merged we update the related files. ESP32C3-RUST board (#74636) depends on this driver and it is important for us to add support for it. @teburd @kartben @jeppenodgaard

File: drivers/sensor/tdk/icm42670/icm42670_i2c.c [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+	if (bank) {
+		res = i2c_read_mreg(bus, address, bank, data, len);
+	} else {
+		res = i2c_burst_read_dt(&bus->i2c, address, data, len);
+	}
+
+	return res;
+}
+
+static inline int i2c_write_mreg(const union icm42670_bus *bus, uint16_t reg, uint8_t bank,
```
Comment (jeppenodgaard) [inline drivers/sensor/tdk/icm42670/icm42670_i2c.c:68]: Why is this inline?
From linux coding style, the inline disease:
> A reasonable rule of thumb is to not put inline at functions that have more than 3 lines of code in them.

File: dts/bindings/sensor/invensense,icm42670-i2c.yaml [hunk 6]
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) 2022 Esco Medical ApS
```
Comment (jeppenodgaard) [inline dts/bindings/sensor/invensense,icm42670-i2c.yaml:1]: It looks like this needs to be updated since it a new file you created.

**Review by jeppenodgaard — COMMENTED** [review]
Are you able to test this with spi?

File: drivers/sensor/tdk/icm42670/icm42670_i2c.c (hunk 5 above)
Comment (LucasTambor) [reply drivers/sensor/tdk/icm42670/icm42670_i2c.c]: Probably inspired by the `icm42670_spi.c` driver that uses `static inline int spi_write_register` but I see your point. Removed from I2C driver.

**Review by teburd — APPROVED** [review]

**Comment by LucasTambor** [comment]
> Are you able to test this with spi?

Unfortunately I am not. But I tried to be as unintrusive as possible, I don't see how the changes could affect the driver's behavior.

**Review by sylvioalves — APPROVED** [review]

**Review by marekmatej — APPROVED** [review]

**Review by jeppenodgaard — APPROVED** [review]

**Pull request merged**